File: solution/160/solution.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None;

    def printval(self):
        print self.val

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        headAset = set()
        while headA:
            headAset.add(headA)
            headA = headA.next
        while headB:
            if headB in headAset:
                return headB
            headB = headB.next
        return None

if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    e = ListNode(5)
    f = ListNode(6)
    g = ListNode(7)
    a.next = b
    b.next = c
    c.next = e
    e.next = f
    g.next = c
    s = Solution()
    result = s.getIntersectionNode(a, g)
    result.printval()

File: solution/160/solution.cpp
#include <stdio.h>
#include <set>

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL){}
};

void printALL(ListNode *head, const char *name){
    printf("%s: ", name);
    if(!head){
        printf("\n");
        return;
    }
    while(head){
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");
}

class Solution{
public:
    ListNode* getIntersectionNode(ListNode *headA, ListNode *headB){
        std::set<ListNode*> listaset;
        while(headA){
            listaset.insert(headA);
            headA = headA->next;
        }
        while(headB){
            if(listaset.find(headB)!=listaset.end()){
                return headB;
            }
            headB = headB->next;
        }
        return NULL;
    }
};

int main(){
    ListNode a(1);
    ListNode b(2);
    ListNode c(3);
    ListNode d(4);
    ListNode e(5);
    ListNode f(6);
    ListNode g(7);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    e.next = &f;
    f.next = &g;
    g.next = &b;
    ListNode *headA = &a;
    ListNode *headB = &e;
    ListNode *result = NULL;
    printALL(headA, "headA");
    printALL(headB, "headB");
    Solution s;
    result = s.getIntersectionNode(headA, headB);
    printALL(result, "result");
}